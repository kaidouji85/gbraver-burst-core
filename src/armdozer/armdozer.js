// @flow

/** アームドーザID */
export type ArmDozerId = string;

// TODO バーストの種類を追加する

/** アームドーザ基本情報 */
export type Armdozer = {
  /** アームドーザID */
  id: ArmDozerId,
  /** 名前 */
  name: string,
  /** 最大HP */
  maxHp: number,
  /** 最大バッテリー */
  maxBattery: number,
  /** 攻撃 */
  power: number,
  /** スピード */
  speed: number,
  /** アームドーザの外見を指定する */
  appearance: string,
};
