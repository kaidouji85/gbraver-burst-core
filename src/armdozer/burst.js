// @flow

/** バースト */
export type Burst = RecoverBattery | BuffPower | Barrier;

/** バッテリー回復 */
export type RecoverBattery = {
  type: 'RecoverBattery',
  /** バッテリー回復料 */
  recoverBattery: number,
};

/** 攻撃力バフ */
export type BuffPower = {
  type: 'BuffPower',
  /** バッテリー回復料 */
  recoverBattery: number,
  /** 攻撃力アップ */
  buffPower: number,
  /** バフ継続ターン数 */
  duration: number
};

/** バリア */
export type Barrier = {
  type: 'Barrier',
  /** バリアに触れた時のダメージ */
  damage: number,
  /** バリア継続ターン */
  duration: number,
};
