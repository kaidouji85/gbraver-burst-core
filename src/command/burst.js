// @flow

/** バーストコマンド */
export type BurstCommand = {
  type: 'BURST_COMMAND'
};