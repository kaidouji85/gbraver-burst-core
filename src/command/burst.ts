import { z } from "zod";

/** バーストコマンド */
export type BurstCommand = {
  type: "BURST_COMMAND";
};

/** バーストコマンド zodスキーマ */
export const BurstCommandSchema = z.object({
  type: z.literal("BURST_COMMAND"),
});

/**
 * 指定したコマンドをバーストコマンドにパースする
 * @param origin パースするコマンド
 * @return パース結果、パースできない場合はnull
 */
export function parseBurstCommand(origin: unknown): BurstCommand | null {
  const result = BurstCommandSchema.safeParse(origin);
  return result.success ? result.data : null;
}
