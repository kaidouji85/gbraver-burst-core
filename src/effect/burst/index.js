// @flow

import type {GameState} from "../../state/game-state";
import type {PlayerId} from "../../player/player";
import type {PlayerState} from "../../state/player-state";
import {recoverBattery} from "./recover-battery";
import {buffPower} from "./buff-power";

/**
 * バーストを実施する
 *
 * @param lastState 最新の状態
 * @param burstPlayerId バーストするプレイヤーID
 * @return バースト実施後の状態
 */
export function burst(lastState: GameState, burstPlayerId: PlayerId): GameState {
  const burstPlayer = lastState.players.find(v => v.playerId === burstPlayerId);
  const otherPlayer = lastState.players.find(v => v.playerId !== burstPlayerId);
  if (!burstPlayer || !otherPlayer) {
    return lastState;
  }

  const updatedPlayers = updateForBurst(burstPlayer, otherPlayer);
  const sortedPlayers = lastState.players
    .map(player => updatedPlayers.find(v => v.playerId === player.playerId) || player);
  return {
    ...lastState,
    players: sortedPlayers,
    effect: {
      name: 'BurstEffect',
      burstPlayer: burstPlayer.playerId,
      burst: burstPlayer.armdozer.burst,
    }
  };
}

/**
 * 各プレイヤーのステータスをバースト実施後のものに更新する
 *
 * @param burstPlayer バーストするプレイヤーの状態
 * @param otherPlayer それ以外のプレイヤーの状態
 * @return バースト実施後の状態
 */
export function updateForBurst(burstPlayer: PlayerState, otherPlayer: PlayerState): PlayerState[] {
  switch (burstPlayer.armdozer.burst.type) {
    case 'RecoverBattery':
      return recoverBattery(burstPlayer, otherPlayer);
    case 'BuffPower':
      return buffPower(burstPlayer, otherPlayer);
    default:
      return [burstPlayer, otherPlayer];
  }
}
