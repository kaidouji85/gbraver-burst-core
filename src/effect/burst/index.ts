import type { PlayerId } from "../../player/player";
import type { GameState, GameStateX } from "../../state/game-state";
import { batteryLimitBreak } from "./battery-limit-break";
import { buffPower } from "./buff-power";
import type { BurstEffect } from "./burst-effect";
import { continuousActivePlayer } from "./continuous-active-player";
import { lightningBarrier } from "./lightning-barrier";
import { recoverBattery } from "./recover-battery";

/**
 * バースト効果を適用する
 * @param lastState 最新状態
 * @param burstPlayerId バーストするプレイヤーID
 * @return 更新結果、更新できない場合は例外を投げる
 */
export function burstEffect(
  lastState: GameState,
  burstPlayerId: PlayerId,
): GameStateX<BurstEffect> {
  const burstPlayer = lastState.players.find(
    (v) => v.playerId === burstPlayerId,
  );
  if (!burstPlayer) {
    throw new Error("not found burst player");
  }

  if (burstPlayer.armdozer.burst.type === "RecoverBattery") {
    return recoverBattery(lastState, burstPlayerId, burstPlayer.armdozer.burst);
  }

  if (burstPlayer.armdozer.burst.type === "BuffPower") {
    return buffPower(lastState, burstPlayerId, burstPlayer.armdozer.burst);
  }

  if (burstPlayer.armdozer.burst.type === "LightningBarrier") {
    return lightningBarrier(
      lastState,
      burstPlayerId,
      burstPlayer.armdozer.burst,
    );
  }

  if (burstPlayer.armdozer.burst.type === "ContinuousAttack") {
    return continuousActivePlayer(
      lastState,
      burstPlayerId,
      burstPlayer.armdozer.burst,
    );
  }

  if (burstPlayer.armdozer.burst.type === "BatteryLimitBreak") {
    return batteryLimitBreak(
      lastState,
      burstPlayerId,
      burstPlayer.armdozer.burst,
    );
  }

  throw new Error("burst not found");
}

/**
 * 指定したプレイヤーのバーストを利用不能にする
 * @param lastState 最新状態
 * @return 更新結果
 */
export function disableBurst(
  lastState: GameStateX<BurstEffect>,
): GameStateX<BurstEffect> {
  const burstPlayer = lastState.players.find(
    (v) => v.playerId === lastState.effect.burstPlayer,
  );

  if (!burstPlayer) {
    throw new Error("not found burst player");
  }

  const updatedBurstPlayer = {
    ...burstPlayer,
    armdozer: { ...burstPlayer.armdozer, enableBurst: false },
  };
  const updatedPlayers = lastState.players.map((v) =>
    v.playerId === updatedBurstPlayer.playerId ? updatedBurstPlayer : v,
  );
  return { ...lastState, players: updatedPlayers };
}

/**
 * バーストを実行する
 * @param lastState 最新の状態
 * @param burstPlayerId バーストするプレイヤーID
 * @return バースト結果、実行不可能な場合はnullを返す
 */
export function burst(
  lastState: GameState,
  burstPlayerId: PlayerId,
): GameStateX<BurstEffect> {
  const doneBurstEffect = burstEffect(lastState, burstPlayerId);
  return disableBurst(doneBurstEffect);
}