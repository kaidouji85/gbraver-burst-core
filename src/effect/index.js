// @flow

import type {InputCommand} from "./input-command/input-command";
import type {Battle} from "./battle/battle";
import type {TurnChange} from "./turn-change/turn-change";
import type {StartGame} from "./start-game/start-game";
import type {BurstEffect} from "./burst/burst-effect";
import type {GameEnd} from "./game-end/game-end";
import type {BatteryDeclaration} from "./battery-declaration/battery-declaration";
import type {Reflect} from "./reflect/reflect";
import type {UpdateRemainingTurn} from "./update-remaning-turn/update-remaining-turn";

/** ゲーム中に発生する効果をまとめたもの */
export type Effect =
  StartGame |
  GameEnd |
  InputCommand |
  Reflect |
  BatteryDeclaration |
  Battle |
  TurnChange |
  BurstEffect |
  UpdateRemainingTurn;
