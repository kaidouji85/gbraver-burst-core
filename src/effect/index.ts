import { z } from "zod";

import { BatteryDeclaration, BatteryDeclarationSchema } from "./battery-declaration/battery-declaration";
import { Battle, BattleSchema } from "./battle/battle";
import { BurstEffect, BurstEffectSchema } from "./burst/burst-effect";
import { GameEnd, GameEndSchema } from "./game-end/game-end";
import { InputCommand, InputCommandSchema } from "./input-command/input-command";
import { PilotSkillEffect, PilotSkillEffectSchema } from "./pilot-skill/pilot-skill-effect";
import { Reflect, ReflectSchema } from "./reflect/reflect";
import { RightItself, RightItselfSchema } from "./right-itself/right-itself";
import { StartGame, StartGameSchema } from "./start-game/start-game";
import { TurnChange, TurnChangeSchema } from "./turn-change/turn-change";
import { UpdateRemainingTurn, UpdateRemainingTurnSchema } from "./update-remaning-turn/update-remaining-turn";

/** ゲーム中に発生する効果をまとめたもの */
export type Effect =
  | StartGame
  | GameEnd
  | InputCommand
  | Reflect
  | BatteryDeclaration
  | Battle
  | TurnChange
  | BurstEffect
  | UpdateRemainingTurn
  | RightItself
  | PilotSkillEffect;

/** Effect zodスキーマ */
export const EffectSchema = z.union([
  StartGameSchema,
  GameEndSchema,
  InputCommandSchema,
  ReflectSchema,
  BatteryDeclarationSchema,
  BattleSchema,
  TurnChangeSchema,
  BurstEffectSchema,
  UpdateRemainingTurnSchema,
  RightItselfSchema,
  PilotSkillEffectSchema,
]);