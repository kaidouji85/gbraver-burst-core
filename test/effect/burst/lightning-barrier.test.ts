import { burst } from "../../../src/effect/burst";
import { EMPTY_ARMDOZER_EFFECT } from "../../../src/empty/amrdozer-effect";
import { EMPTY_ARMDOZER_STATE } from "../../../src/empty/armdozer";
import { EMPTY_GAME_STATE } from "../../../src/empty/game-state";
import { EMPTY_PLAYER_STATE } from "../../../src/empty/player";
import { GameState } from "../../../src/state/game-state";
import { PlayerState } from "../../../src/state/player-state";

/** バースト発動者 */
const burstPlayer: PlayerState = {
  ...EMPTY_PLAYER_STATE,
  playerId: "burst",
  armdozer: {
    ...EMPTY_ARMDOZER_STATE,
    enableBurst: true,
    battery: 1,
    maxBattery: 5,
    effects: [EMPTY_ARMDOZER_EFFECT, EMPTY_ARMDOZER_EFFECT],
    burst: {
      type: "LightningBarrier",
      damage: 1000,
      duration: 2,
      recoverBattery: 3,
    },
  },
};

/** それ以外のプレイヤー */
const otherPlayer: PlayerState = { ...EMPTY_PLAYER_STATE, playerId: "other" };

/** ゲームステート */
const lastState: GameState = {
  ...EMPTY_GAME_STATE,
  activePlayerId: otherPlayer.playerId,
  players: [otherPlayer, burstPlayer],
};

test("電撃バリアバーストの適用が正しくできる", () => {
  const result = burst(lastState, burstPlayer.playerId);
  expect(result).toMatchSnapshot();
});
