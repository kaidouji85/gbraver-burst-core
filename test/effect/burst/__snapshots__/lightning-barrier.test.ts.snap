// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`電撃バリアバーストの適用が正しくできる 1`] = `
{
  "activePlayerId": "",
  "effect": {
    "burst": {
      "damage": 1000,
      "duration": 2,
      "recoverBattery": 3,
      "type": "LightningBarrier",
    },
    "burstPlayer": "burst",
    "name": "BurstEffect",
  },
  "players": [
    {
      "armdozer": {
        "battery": 5,
        "burst": {
          "recoverBattery": 5,
          "turnStartBatteryCorrect": 1,
          "type": "RecoverBattery",
        },
        "effects": [],
        "enableBurst": true,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": true,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 2,
          "type": "RecoverBatterySkill",
        },
      },
      "playerId": "other",
    },
    {
      "armdozer": {
        "battery": 4,
        "burst": {
          "damage": 1000,
          "duration": 2,
          "recoverBattery": 3,
          "type": "LightningBarrier",
        },
        "effects": [
          {
            "period": {
              "remainingTurn": 1,
              "type": "TurnLimit",
            },
            "type": "Empty",
          },
          {
            "period": {
              "remainingTurn": 1,
              "type": "TurnLimit",
            },
            "type": "Empty",
          },
          {
            "damage": 1000,
            "effect": "Lightning",
            "period": {
              "remainingTurn": 2,
              "type": "TurnLimit",
            },
            "type": "TryReflect",
          },
        ],
        "enableBurst": false,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": true,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 2,
          "type": "RecoverBatterySkill",
        },
      },
      "playerId": "burst",
    },
  ],
}
`;
